### PR TITLE
Add user resolver based on user's resident org traversing to ansectors

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Part of fix https://github.com/wso2/product-is/issues/14154

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/constant/SQLConstants.java [hunk 1]
```diff
@@ -27,13 +27,18 @@
...
 
     public static final String PERMISSION_LIST_PLACEHOLDER = "_PERMISSION_LIST_";
-    public static final String GET_PERMISSIONS_FOR_USER = "SELECT UM_RESOURCE_ID FROM UM_ORG_ROLE_USER INNER JOIN " +
-            "UM_ORG_ROLE ON UM_ORG_ROLE_USER.UM_ROLE_ID = UM_ORG_ROLE.UM_ROLE_ID INNER JOIN " +
-            "UM_ORG_ROLE_PERMISSION ON UM_ORG_ROLE_USER.UM_ROLE_ID = UM_ORG_ROLE_PERMISSION.UM_ROLE_ID INNER JOIN " +
-            "UM_ORG_PERMISSION ON UM_ORG_ROLE_PERMISSION.UM_PERMISSION_ID = UM_ORG_PERMISSION.UM_ID WHERE " +
-            "UM_ORG_ROLE_USER.UM_USER_ID = :" + DB_SCHEMA_COLUMN_USER_ID + "; AND UM_ORG_ROLE.UM_ORG_ID = :" +
-            DB_SCHEMA_COLUMN_ORGANIZATION_ID + ";";
-
+    public static final String CHECK_USER_HAS_PERMISSION_TO_ORG_BY_DIRECT_ROLES =
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/constant/SQLConstants.java:30]: what is meant by direct role here

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Service to resolve user's resident organization.
+ */
+public interface OrganizationUserResidentResolverService {
+
+    /**
+     * Resolve the user from resident organization.
+     *
+     * @param userName               Username.
+     * @param accessedOrganizationId Organization that user trying to access.
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java:35]:
```suggestion
     * @param accessedOrganizationId Organization that user is trying to access.
```

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     * @return User object based on user resident.
+     * @throws OrganizationManagementException Error occurred while resolving resident org of the user.
+     */
+    Optional<User> resolveUserFromResidentOrgByUsernameAndAccessedOrg(String userName,
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java:39]: IMO we can remove `ByUsernameAndAccessedOrg` part from the method name as it is understandable from the method params

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    @Override
+    public Optional<User> resolveUserFromResidentOrgByUsernameAndAccessedOrg(String userName,
+                                                                             String accessedOrganizationId)
+            throws OrganizationManagementException {
+
+        User resolvedUser = null;
+        try {
+            List<String> ancestorOrganizationIds =
+                    organizationManagementDAO.getAncestorOrganizationIds(accessedOrganizationId);
+            if (ancestorOrganizationIds != null) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverServiceImpl.java:56]: shall we use !CollectionUtils.isEmpty

File: components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/constant/SQLConstants.java (hunk 1 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/constant/SQLConstants.java:30]: User assigned to a role directly (not getting permissions through  group assignment to the role)
 If you think this variable name is not clear what about `CHECK_USER_HAS_PERMISSION_TO_ORG_THROUGH_USER_ROLE_ASSIGNMENT`

Comment (dewniMW) [reply components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/constant/SQLConstants.java:30]: how about CHECK_USER_HAS_PERMISSION_TO_ORG_THROUGH_ROLE_ASSIGNMENT

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/constant/SQLConstants.java]: fixed in [38bc99c](https://github.com/wso2-extensions/identity-organization-management/pull/92/commits/38bc99cf391a2975a4092b77e0ab7c75bde09653)

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java (hunk 3 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java]: fixed in [38bc99c](https://github.com/wso2-extensions/identity-organization-management/pull/92/commits/38bc99cf391a2975a4092b77e0ab7c75bde09653)

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java (hunk 2 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverService.java]: fixed in [38bc99c](https://github.com/wso2-extensions/identity-organization-management/pull/92/commits/38bc99cf391a2975a4092b77e0ab7c75bde09653)

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                            if (userHasAccessPermissions) {
+                                resolvedUser = user;
+                            }
+                        }
+                    }
+                }
+            }
+        } catch (UserStoreException | OrganizationManagementAuthzServiceServerException e) {
+            throw new OrganizationManagementException(
+                    ERROR_CODE_ERROR_WHILE_RESOLVING_USER_FROM_RESIDENT_ORG.getMessage(),
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationUserResidentResolverServiceImpl.java:77]: is the error description not required here?

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**